### PR TITLE
[Inference API] Extract optional long instead of integer in RateLimitSettings#of

### PR DESCRIPTION
**Pull request by timgrein**
`RateLimitSettings` expects a `long` in the constructor, but in [`RateLimitSettings#of`](https://github.com/elastic/elasticsearch/blob/77ce60530ceb562920bd0ccaa424f1ec58ce781b/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/settings/RateLimitSettings.java#L35) we read the number using `extractOptionalPositiveInteger`. If you set `requests_per_minute` to a number, which doesn't fit into an Integer you'll run into an error.

Example:
```
{
    "service": "azureopenai",
    "service_settings": {
        "resource_name":"test",
        "deployment_id": "gpt-4",
        "api_version": "2024-02-01",
        "api_key": "{{AZURE_OPEN_AI_API_KEY}}",
        "rate_limit": {
            "requests_per_minute": 10000000000
        }
    }
}
```

Therefore I've introduced a new method `extractOptionalPositiveLong`, which extracts an optional `Long`. I've made it transparent to the callers that under the hood integers can be extracted, too. This removes the need of callers of `new RateLimitSettings(...)` to always specify that it's a long, otherwise `new RateLimitSettings(3)` would provoke an error on reading the value back.

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java [hunk 1]
```diff
@@ -91,6 +92,40 @@ public static <T> T removeAsType(Map<String, Object> sourceMap, String key, Clas
...
+    /**
+     * Remove the object from the map and cast to one of the expected types.
+     * If the object cannot be cast to one of the types an error is added to the
+     * {@code validationException} parameter
+     *
+     * @param sourceMap Map containing fields
+     * @param key The key of the object to remove
+     * @param types One of the expected types of the removed object
+     * @param validationException If the value is not of type {@code type}
+     * @return {@code null} if not present else the object cast to type T
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java:104]:
```suggestion
     * @return {@code null} if not present else the object cast to the first assignable type in the {@types} list
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java [hunk 2]
```diff
@@ -91,6 +92,40 @@ public static <T> T removeAsType(Map<String, Object> sourceMap, String key, Clas
         }
     }
 
+    /**
+     * Remove the object from the map and cast to one of the expected types.
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java:96]: Editing the comment to make it clear that the order of the types in `types` is important.

```suggestion
     * Remove the object from the map and cast to first assignable type in the expected types list.
```

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java [hunk 3]
```diff
@@ -91,6 +92,40 @@ public static <T> T removeAsType(Map<String, Object> sourceMap, String key, Clas
...
+     * If the object cannot be cast to one of the types an error is added to the
+     * {@code validationException} parameter
+     *
+     * @param sourceMap Map containing fields
+     * @param key The key of the object to remove
+     * @param types One of the expected types of the removed object
```
Comment (davidkyle) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java:102]:
```suggestion
     * @param types The expected types of the removed object
```

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java (hunk 2 above)
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/ServiceUtils.java]: Makes a lot of sense, thanks!

**Comment by timgrein** [comment]
@elasticmachine update branch

**Comment by timgrein** [comment]
@elasticmachine run elasticsearch-ci

**Comment by timgrein** [comment]
@elasticmachine update branch

**Comment by timgrein** [comment]
@elasticmachine update branch

**Comment by timgrein** [comment]
@elasticmachine update branch

**Pull request merged**